server/src/main/java/org/elasticsearch/index/mapper/SourceToParse.java

```diff
@@ -40,14 +40,15 @@ public static SourceToParse source(String index, String type, String id, BytesRe
 
     private final String id;
 
-    private String routing;
+    private final String routing;
 
-    private XContentType xContentType;
+    private final XContentType xContentType;
 
-    private SourceToParse(String index, String type, String id, BytesReference source, XContentType xContentType) {
+    public SourceToParse(String index, String type, String id, String routing, BytesReference source, XContentType xContentType) {
```

Comment:
If we want to avoid sprinkling `null` all over the codebase, we can also have a second constructor here that does not take `routing` as parameter and simply delegates to this constructor here with routing set to `null`.